### DISCUSSION TITLE
Redability report generator feautures

### DISCUSSION THREAD

**Discussion by Vilhelm-Ian** [mortii/anki-morphs#114] [Ideas]
This is a list of things I want to add to the redability report generator. I would like to eventually to add them, and not bother you mortii to do the hard work.

1.  Calculate the percentage based on the number of occurences of a word. Like in the following sentence with ten words.
"I am too shy to speak to that many people." If I only knew the word "to" the percentage stat would be 20% since to appears twice and there are 10 words.

2. Compare only the base form of words. In the following sentence "She borrowed my favorite book but returned two different books. " If I only knew the word book I would get 20% since book and books appeear and there are 10 words.

3. In the absolute numbers ability to click on unknown morphs and to get a list of unknown morphs.

4. (Optional) Ability to add a percentage cut-off. Like if I wanted to get a 95% comprehension it would generate a different amount of absolute unknown morphs. So when you click you wil get which words you need to know to reach that level.

**Comment by mortii** [comment]
Cool!

> 1. Calculate the percentage based on the number of occurences of a word. Like in the following sentence with ten words.
"I am too shy to speak to that many people." If I only knew the word "to" the percentage stat would be 20% since to appears twice and there are 10 words.

I thought it already did that, but I never checked. That would be a good improvement.

> 2. Compare only the base form of words. In the following sentence "She borrowed my favorite book but returned two different books. " If I only knew the word book I would get 20% since book and books appeear and there are 10 words.

I'm not sure if that would be a good idea. If you have a sentence like:
```
She was gone last week, he went shortly after, and I go tomorrow.
```
If I only know 'go', but not 'went', nor 'gone', then the stats would be misleading....

> 3. In the absolute numbers ability to click on unknown morphs and to get a list of unknown morphs.

Great idea. How would the list be displayed? Like the 'view morphemes' option in the browser?

> 4. (Optional) Ability to add a percentage cut-off. Like if I wanted to get a 95% comprehension it would generate a different amount of absolute unknown morphs. So when you click you wil get which words you need to know to reach that level.

I don't know how I feel about that. It was the least understandable option in the morphman readability analyzer. It might cross over into the frequency file generator territory.

**Comment by mortii** [comment]
> > 4. (Optional) Ability to add a percentage cut-off. Like if I wanted to get a 95% comprehension it would generate a different amount of absolute unknown morphs. So when you click you wil get which words you need to know to reach that level.
> 
> I don't know how I feel about that. It was the least understandable option in the morphman readability analyzer. It might cross over into the frequency file generator territory.

I think I have a semi-genius idea: If you click on the list of unknown morphs, the list is ordered by frequency, and in the margin it shows the percentile of how much you much you would know if you reach that morph? Does that make sense?

**Reply by Vilhelm-Ian** [reply]
yes that is really smart. How did you think of it?

**Reply by mortii** [reply]
just tried to think some other alternative to having a comprehension option, and that seemed like it would accomplish the same thing, and maybe even more.

**Comment by mortii** [comment]
> 1. Calculate the percentage based on the number of occurences of a word. Like in the following sentence with ten words.
"I am too shy to speak to that many people." If I only knew the word "to" the percentage stat would be 20% since to appears twice and there are 10 words.

@Vilhelm-Ian I think I updated the morphemizers to return morphs in sets rather than lists, which might have contributed to this problem. Feel free to change it back to lists if necessary.

**Comment by mortii** [comment]
@Vilhelm-Ian I'm basically done with everything else so I can help with this now. How far have you gotten?

**Reply by Vilhelm-Ian** [reply]
I did nothing. I shall start right now after sending this message.

**Reply by mortii** [reply]
@Vilhelm-Ian that's completely fine. I actually have some ideas that I want to go through, give me a sec.

**Comment by mortii** [comment]
Roadmap:

1.  Use lists instead of sets for storing morphs, that way we get more accurate stats. I could probably do this in ~20 minutes
2.  Add detailed reports for each entry

I'm thinking we could have a column all the way to the right in this table:
![readability-report-generator-percentages](https://github.com/mortii/anki-morphs/assets/15674619/89acb318-8bb6-4455-b0d5-77f6f766de24)

that has buttons that say 'detailed report' or something like that. When you click it you open a new dialog. That dialog can have have this information in separate tabs:

### tab 1: fine grained morph stats


``` text
percentile |  morph-inflection | morph-lemma |learning status 
0%         |   the             |  the        | known
2%         |   a               |  a          | learning
``` 

### tab 2: frequency file
have the raw frequency file here that can be copy pasted. This way one wouldn't have to go through the trouble of doing that with the frequency file generator. This would kinda replace the study plan feature in morphman.

thoughts @Vilhelm-Ian ?

**Reply by Vilhelm-Ian** [reply]
I like both ideas
I shall add them.

Actually I don't like the second idea. The thing people want to know about with 95% cut off is how many more morphs they need to learn to get to a certain percentaile I thought you meant a button where you get a list of all the unknown morphs ordered by frequency. And next to the morph what percentile you would reach if you knew all the unknown morphs below it.

**Reply by Vilhelm-Ian** [reply]
the third option. I think it's nice. But it's not fesabile with what I said.

How about this. The button "detailed-report" will open a window with two buttons. By default will show what you said and people can copy paste it as a frequency file.

Second button will do what I said and people can see how many unknown morphs they roughly need to get to a certain percentage

**Reply by mortii** [reply]
> Actually I don't like the second idea. The thing people want to know about with 95% cut off is how many more morphs they need to learn to get to a certain percentaile I thought you meant a button where you get a list of all the unknown morphs ordered by frequency. And next to the morph what percentile you would reach if you knew all the unknown morphs below it.

that is exactly what it is, except it also includes every morph, also known ones. I forgot to add that I think a we could have a checkbox that filters out every morph that is not unknown, so you would be left with just a table like you describe. It's kinda hard to illustrate without making it, but I could try drawing it or something if you want.

> How about this. The button "detailed-report" will open a window with two buttons. By default will show what you said and people can copy paste it as a frequency file.
>
>Second button will do what I said and people can see how many unknown morphs they roughly need to get to a certain percentage

I was thinking it could be similar to the  bottom of the current readability report generator:

![Screenshot from 2024-01-18 18-52-25](https://github.com/mortii/anki-morphs/assets/15674619/761643ce-f931-4ef1-aae9-1c840f044f43)

Absolute numbers -> Morph Stats
Percentages -> Frequency File

**Reply by Vilhelm-Ian** [reply]
@mortii  I get it now.  I like the idea. Now it makes sense.

**Comment by Vilhelm-Ian** [comment]
https://github.com/Vilhelm-Ian/anki-morphs/commit/ce2735171be4deef3fc3195b035b03dda9d448b2


do you like the approach I took?
Mortii can you leave this one to me. I really want to contribute

**Reply by Vilhelm-Ian** [reply]
I get better results
![image](https://github.com/mortii/anki-morphs/assets/81630959/77f131ee-2ee6-4889-998e-e0b1371b8473)
here is the absolute table
![image](https://github.com/mortii/anki-morphs/assets/81630959/d2d7eb8a-4772-4253-a9e8-981be74dc287)
the old way the relative table would have told me that I know 17% of the morphs

**Reply by mortii** [reply]
> do you like the approach I took?

It looks ok

> Mortii can you leave this one to me. I really want to contribute

Which parts do you want to do?

**Comment by Vilhelm-Ian** [comment]
> Which parts do you want to do?

Well I already fixed the way they are calculated. All that is left to make the the list of morphs and percentile. Am I supposed to render the list as text so they can copy it or as a table and a button to export?

**Reply by mortii** [reply]
> Well I already fixed the way they are calculated.

Nice. I was thinking that this is essentially a bug fix and once this is done I can release V1 and stop being in alpha. Could you create a pull request with those changes first? After that you work on the rest at your own pace, no stress.

> Am I supposed to render the list as text so they can copy it or as a table and a button to export?

That is actually a great point, an export button would save having to copy and paste the frequency file. Maybe just do that instead of having it in it's own tab. 

The morph stats definitely need to be in a QTableWidget.
